### ISSUE TITLE
Prisma Client isn't generated on deploying to Vercel

### ISSUE THREAD

**Issue by ryands17** [prisma/prisma#5175]
## Bug description

As per the following discussion:

https://github.com/prisma/prisma/discussions/5140

### Deploying prisma to Vercel breaks

So I've started my project from the [official example found here](https://github.com/prisma/prisma-examples/tree/latest/deployment-platforms/vercel-graphql). This project works perfectly when deployed to Vercel directly because it forces Vercel to download, build and generate the prisma client.

Unfortunately, if you were to upload your project to Github and set up auto-deploys on commit, you will soon find a bunch of errors related to `@prisma/client` missing exports in your build outputs. For me these were missing types for my data and even native exports like `empty` and `sql`.

If you were to try to access your API you would most likely be able to see something like `2021-01-17T16:41:33.659Z	undefined	ERROR	Error: You most likely forgot to initialize the Prisma Client. Please run `prisma generate` and try to run it again.` in your function logs.

### Why this happens

Vercel, like many other serverless hosts, uses a `package-lock`-based cache. What this means in practice is:

1. Change is made to the Github
2. Vercel sees the changes
3. Checks if anything has changed in `package-lock`
4. If it's the same as the previous commit, it downloads a cache of `node_modules`

**Unfortunately**, `node_modules` is where the prisma client (which contains all of our types) is kept. And even more unfortunately, changes to our API which would be reflected in our types when running `npm run generate` does not make changes to `package-lock`.

In short this means that even though our app may have gone through significant changes, if `package-lock` isn't deleted manually, our types will not be generated, which in turn will most likely break our API.

## Expected behavior

A workaround should be possible to generate the `PrismaClient` in Vercel.

**Comment by creativiii** [comment]
Thank you for creating an issue. Honestly I think a lot of the problems I had could be fixable by:

- Updating he example (which still has code that refers to Now as well as legacy settings
- Rewriting the guide with a bit more information

Let me know if that's something you're looking for, I'd love to submit a PR with my example and guide edits.

**Comment by 2color** [comment]
I've been able to reproduce this problem.

The essence of the problem seems to be that Vercel ships a cached version of `node_modules` with the functions and in _some_ situations the cached `node_modules` doesn't contain the generated Prisma Client (from the `postinstall` script in Prisma Client or the `build` script of the actual project). There are certain circumstances that cause this cache to get busted, like updating the package-lock.json. 

This is a problem when making changes to the Prisma schema. For example, when you add a new field, migrate the database, and update one of your functions to use the new field. When the changes are committed and pushed, Vercel triggers a build that loads the `node_modules` folder from cache. 


I created a reproduction here: https://github.com/2color/vercel-prisma-reproduction

**Comment by 2color** [comment]
@creativiii It would be a good idea to update the Vercel deployment and change it to use the GitHub flow rather than the CLI. 

I'd be happy to help review a PR. Probably makes sense to find a solution to this problem.

**Comment by creativiii** [comment]
> change it to use the GitHub flow

You mean connecting the repo to Vercel and setting up automatic deployments?

**Comment by 2color** [comment]
> You mean connecting the repo to Vercel and setting up automatic deployments?

Yes

**Comment by creativiii** [comment]
I can do that then, expect a PR soon 😀

**Comment by 2color** [comment]
I was able to reproduce this problem and solve it by adding the `postinstall: prisma generate` to `package.json`.

This was confirmed by the Vercel team who clarified that:
> Having `prisma generate` as the `postinstall` command is a way to ensure that is **always run after the initial build step**, regardless of whether it used the cached node_modules or not.


From my tests, it seems that [the trick that we do when building on Vercel](https://github.com/prisma/prisma/blob/master/src/packages/cli/scripts/install.js#L9), doesn't always actually work (CC: @timsuchanek).


@creativiii I'd suggest you add the `postinstall` script and see if that solves the problem for you.

**Comment by creativiii** [comment]
> I was able to reproduce this problem and solve it by adding the `postinstall: prisma generate` to `package.json`.
> 
> This was confirmed by the Vercel team who clarified that:
> 
> > Having `prisma generate` as the `postinstall` command is a way to ensure that is **always run after the initial build step**, regardless of whether it used the cached node_modules or not.
> 
> From my tests, it seems that [the trick that we do when building on Vercel](https://github.com/prisma/prisma/blob/master/src/packages/cli/scripts/install.js#L9), doesn't always actually work (CC: @timsuchanek).
> 
> @creativiii I'd suggest you add the `postinstall` script and see if that solves the problem for you.

Okay so I've tested this in [this branch](https://github.com/coloredcat/serverless-prisma-vercel-boilerplate/tree/bugfix/prisma-client) and unless I'm missing something I don't think it's working.

My process:

- Create branch and ensured it got deployed on Vercel so that node_modules would be cached
- Changed my schema and added a new value `banned` to the users table
- Added `t.model.banned()` to my User types.
- Deployed

On build I got:

```
10:39:51.493 | NowBuildError: api/_lib/mutation.ts(28,13): error TS2741: Property 'banned' is missing in type '{ username: string; email: string; role: string; password: string; }' but required in type 'UserCreateInput'.
-- | --
10:39:51.493 | at createTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99067:16)
10:39:51.493 | at reportTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99073:23)
10:39:51.493 | at getOutputTypeCheck (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99160:17)
10:39:51.493 | at compile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99238:40)
10:39:51.493 | at compileTypeScript (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98515:31)
10:39:51.493 | at Job.readFile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98541:30)
10:39:51.493 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48057:33)
10:39:51.494 | at /vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48095:32
10:39:51.494 | at Array.map (<anonymous>)
10:39:51.494 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48075:26) {
```

Which is the same error I got before.

Someone suggested I move the default place the prisma client gets generated and I think that might be a better solution for this specific use case? https://www.prisma.io/docs/concepts/components/prisma-client/generating-prisma-client#specifying-a-custom-output-path

**Comment by 2color** [comment]
> Okay so I've tested this in this branch and unless I'm missing something I don't think it's working.

@creativiii It seems like you're hitting a real type error rather than a problem with code generation caching.

Since you added a required `banned` field to your Prisma schema, the following raises a type error:

https://github.com/coloredcat/serverless-prisma-vercel-boilerplate/blob/a75948496341b4cf081c7d5e2490b84d3de776aa/api/_lib/mutation.ts#L26-L34

Because it's missing the `banned` property.

This is also the reason why it's failing in the GitHub action too: https://github.com/coloredcat/serverless-prisma-vercel-boilerplate/runs/1748086274?check_suite_focus=true#step:5:27.

Try adding the `banned` to that `prisma.user.create` call. Alternatively, make the field option or set a default value in the Prisma schema.

Let me know if that works.

**Comment by creativiii** [comment]
> > Okay so I've tested this in this branch and unless I'm missing something I don't think it's working.
> 
> @creativiii It seems like you're hitting a real type error rather than a problem with code generation caching.
> 
> Since you added a required `banned` field to your Prisma schema, the following raises a type error:
> 
> https://github.com/coloredcat/serverless-prisma-vercel-boilerplate/blob/a75948496341b4cf081c7d5e2490b84d3de776aa/api/_lib/mutation.ts#L26-L34
> 
> Because it's missing the `banned` property.
> 
> This is also the reason why it's failing in the GitHub action too: https://github.com/coloredcat/serverless-prisma-vercel-boilerplate/runs/1748086274?check_suite_focus=true#step:5:27.
> 
> Try adding the `banned` to that `prisma.user.create` call. Alternatively, make the field option or set a default value in the Prisma schema.
> 
> Let me know if that works.

You were absolutely correct, that was my mistake. The postinstall script works 👍

**Comment by 2color** [comment]
Closing this issue.

If you encounter this error, ensure that you have `prisma generate` in the `postinstall` hook. This is because the `postinstall` hook runs after the `node_modules` is loaded from cache on Vercel.

**Comment by mattslight** [comment]
This really should be more prominent in the install config, docs or picked up on vercel when using Next.js. Just spent 48 hours blaming next-auth for not returning updated fields

**Comment by dcartertwo** [comment]
+1 just ran into this issue.

**Comment by marcosrjjunior** [comment]
I had a similar issue trying to deploy to vercel using their "output API" a few months ago, [building using github-action and deploying to vercel.](https://marcosrjjunior.com/blog/deploying-nextjs-prisma-using-vercel-output-api)

I could solve the issue by setting the `binaryTargets` on prisma.

```java
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
}
```

but looks like the postinstall script works too, I haven't tested.

**Comment by Aghiles97** [comment]
Adding  "postinstall": "prisma generate"
to my package.json into the scripts :{} section, SOLVED my problem.
Thank you for recommending this.

**Comment by sch-28** [comment]
Hey, 
I tried the solutions provided here, but It didnt fix my issue. Vercel always returns following error:
```
Cannot find module '@prisma/client'
Require stack:
- /var/task/.svelte-kit/output/server/chunks/prisma.js
Did you forget to add it to "dependencies" in `package.json`?
```

**Comment by sch-28** [comment]
I found a solution myself, the issue was caused by how I initialized the prisma client.
I used this:
```
import type { PrismaClient as ImportedPrismaClient } from '@prisma/client';
import { createRequire } from 'module';

const require = createRequire(import.meta.url ?? __filename);
const { PrismaClient: RequiredPrismaClient } = require('@prisma/client');
const _PrismaClient: typeof ImportedPrismaClient = RequiredPrismaClient;
class PrismaClient extends _PrismaClient {}

export const prisma = new PrismaClient();
```

changing it to:
 ```
import { PrismaClient } from '@prisma/client';
export const prisma = new PrismaClient();
```
 for deployment fixed this issue.

**Comment by the-frenchies-spots** [comment]
`<html>
<body>


Hello, my post install works, prisma has been generate. But not found in the build process ? 


Prisma schema loaded from prisma/schema.prisma
--
12:09:19.590 |  
12:09:19.590 | ✔ Generated Prisma Client (4.16.2 \| library) to ./node_modules/.pnpm/@prisma+client@4.16.2_prisma@4.16.2/node_modules/@prisma/client in 256ms
12:09:19.591 |  
12:09:19.591 | ✔ Generated Prisma Client (4.16.2 \| library) to ./node_modules/.pnpm/@prisma+client@4.16.2_prisma@4.16.2/node_modules/@prisma/client in 250ms
12:09:19.591 | You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
12:09:19.591 | ```
12:09:19.591 | import { PrismaClient } from '@prisma/client'
12:09:19.591 | const prisma = new PrismaClient()
12:09:19.591 | ```
12:09:19.626 | Done in 22.5s
12:09:19.942 | Using TypeScript 5.2.2 (local user-provided)
12:09:23.012 | src/service/prisma.service.ts(2,10): error TS2305: Module '"@prisma/client"' has no exported member 'PrismaClient'.
12:09:23.013 | src/service/prisma.service.ts(7,16): error TS2339: Property '$connect' does not exist on type 'PrismaService'.
12:09:23.013 | src/service/prisma.service.ts(11,10): error TS2339: Property '$on' does not exist on type 'PrismaService'.


</body>
</html>`

**Comment by janpio** [comment]
That looks like a bug. Please open a new issue or discussion about this @the-frenchies-spots and fill the template. Thanks.

**Comment by vivienneobrien** [comment]
I still can not get prisma to work with supabase despite following the instructions above/ gone through this page reading it extremely carefully multiple times 🥲 and have not found a solution : 
https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/vercel-caching-issue

```
sh: line 1: prisma: command not found
npm ERR! code 127
npm ERR! path /vercel/path0
npm ERR! command failed
npm ERR! command sh -c prisma generate
npm ERR! A complete log of this run can be found in: /vercel/.npm/_logs/2024-06-22T01_20_33_890Z-debug-0.log
Error: Command "npm install" exited with 127
```

I have also followed these instructions and still nothing:
https://www.prisma.io/docs/orm/prisma-client/deployment/serverless/deploy-to-vercel#:~:text=If%20you%20see%20prisma%3A%20command,to%20be%20a%20standard%20dependency.

**Comment by styxlab** [comment]
Still having this issue when using output directory and deploying to vercel, postinstall did not help:
```
Form processing error: Error [PrismaClientKnownRequestError]: 
Invalid `prisma.user.upsert()` invocation:


ENOENT: no such file or directory, open '/var/task/app/generated/prisma/client/query_compiler_bg.wasm'
    at os.handleRequestError (.next/server/app/(company)/kontakt/page.js:1382:7989)
    at os.handleAndLogRequestError (.next/server/app/(company)/kontakt/page.js:1382:7049)
    at os.request (.next/server/app/(company)/kontakt/page.js:1382:6756)
    at async i (.next/server/app/(company)/kontakt/page.js:1394:6321)
    at async nm (.next/server/app/(company)/kontakt/page.js:1308:22426)
    at async (.next/server/app/(company)/kontakt/page.js:1308:24009)
    at async nf.email (.next/server/app/(company)/kontakt/page.js:1308:23441) {
  code: 'ENOENT',
  clientVersion: '6.10.1',
  meta: [Object]
}
```
